### PR TITLE
Issue#609 Add Deviations to Box Manifests

### PR DESCRIPTION
**Pull request by Gbarra9**
This PR addresses issue [#609](https://github.com/episphere/connect/issues/609).

**Changes related to the issue are as followed:**

Created a deviation Type column list for individual box manifests in 

- Shipping dashboard's packaging 
-Reports Shipping reports
- BPTL's Packages in Transit Page

Other things Changed
- Centered checkbox button in Shipping dashboard's Select one or more boxes to ship table
- Renamed BPTL package in transits's exported function filteredShipped to getRecentBoxesShippedBySiteNotReceived
- Renamed previous searchSpecimenInstitute function to filterSpecimenCollectionList
- Removed GET request logic from previous searchSpecimenInstitute function and created outside function with GET request

### CODE REVIEW

File: src/shared.js [hunk 1]
```diff
@@ -834,21 +834,31 @@ export const removeBag = async(boxId, bags) => {
...
 
 /**
- * Fetches biospecimen collection data from the database, and removes '0008', '0009' and deviation tubes from each collection
- * @returns {Array} List of biospecimen collections
+ * Fetches biospecimen collection data from the database
+ * @returns {Promise}
  */
-export const searchSpecimenInstitute = async () => {
-    //https://us-central1-nih-nci-dceg-connect-dev.cloudfunctions.net/biospecimen?api=searchSpecimen
+export const searchSpecimenInstitute = async() => { // rename to searchSpecimenInstitute
```
Comment (anthonypetersen) [inline src/shared.js:840]: comment needed?

File: src/shared.js [hunk 2]
```diff
@@ -834,21 +834,31 @@ export const removeBag = async(boxId, bags) => {
...
+}
+
+/**
+ * Fetches biospecimen collection data from the database, and removes '0008', '0009' and deviation tubes from each collection
+ * @returns {Array} List of biospecimen collections
+ */
+export const filterSpecimenCollectionList = async () => {
+    //https://us-central1-nih-nci-dceg-connect-dev.cloudfunctions.net/biospecimen?api=searchSpecimen
+    const searchSpecimenInstituteResponse = await searchSpecimenInstitute()
+    const searchSpecimenInstituteList = searchSpecimenInstituteResponse?.data ? searchSpecimenInstituteResponse.data : []
```
Comment (anthonypetersen) [inline src/shared.js:858]: I'm assuming this is your method of error handling a 400 code / error resulting from searchSpecimenInstitute() ?

File: src/events.js [hunk 3]
```diff
@@ -3642,7 +3665,28 @@ export const populateReportManifestTable = (currPage) => {
...
+
+            if (currTube) {
+                const acceptedDeviationList = getSpecimenDeviation(searchSpecimenInstituteList, currTube);
+                let deviationString = '';
+                if (acceptedDeviationList.length === 1) {
+                    for (const deviationLabel of acceptedDeviationList) {
+                        deviationString = `${deviationLabel} <br><br>`;
+                    }
+                    currRow.insertCell(3).innerHTML = deviationString;
+                }
```
Comment (anthonypetersen) [inline src/events.js:3677]: see comment above about combining IF / ELSE IF

File: src/events.js [hunk 4]
```diff
@@ -3782,4 +3826,47 @@ const findScannedIdInBoxesNotShippedObject = (getAllBoxesWithoutConversionRespon
...
+    
+    // Flatten array to single object 
+    const specimenObj = (specimenObjList.length) ? Object.assign(...specimenObjList) : {};
+    const acceptedDeviationArr = [];
+
+    for (const key in specimenObj) {
+        const currSpecimenKey = specimenObj[key]
+        // loop over all properties to find scannedId property - 825582494 
+        if (!currSpecimenKey.hasOwnProperty(scannedId)) continue;
+        if (currSpecimenKey[scannedId] === currTube && currSpecimenKey[isCollected] === conceptIds.yes && currSpecimenKey[isDeviated] === conceptIds.yes) {
```
Comment (anthonypetersen) [inline src/events.js:3855]: could add conditionality from line 3854 into line 3855..

if( currSpecimenKey[scannedId] && .... )

**Review by anthonypetersen — APPROVED** [review]
looks good... most of my stuff is just suggestions / ideas

File: src/shared.js (hunk 1 above)
Comment (Gbarra9) [reply src/shared.js:840]: Thanks for spotting this.

File: src/tubeValidation.js [hunk 5]
```diff
@@ -35,6 +35,33 @@ const tubeNotCollectedReason5 = {'concept': 889386523, 'label': 'Supply Unavaila
 const tubeNotCollectedOptions1 = [tubeNotCollectedReason1, tubeNotCollectedReason2, tubeNotCollectedReason3, tubeNotCollectedReason5, tubeNotCollectedReason4];
 const tubeNotCollectedOptions2 = [tubeNotCollectedReason2, tubeNotCollectedReason3, tubeNotCollectedReason5, tubeNotCollectedReason4];
 
+export const refusedShippingDeviationConceptList = [deviationReason1.concept, deviationReason12.concept,deviationReason15.concept, deviationReason17.concept, deviationReason20.concept]
+
+export const deviationReasons = [
```
[the remark this answers is not in the document]
Comment (we-ai) [reply src/tubeValidation.js:40]: For clarity and consistance, this variable can be the single source of deviations and replace variables in lines 1-22. For example `const deviationReasons = {deviationReason1:{'concept': 472864016, 'label':'Broken (discard do not ship)'}, deviationReason2:{...}, ...}`. If needed, you can generate an array on the fly (eg `Object.keys(deviationReasons)`).

File: src/shared.js (hunk 2 above)
Comment (Gbarra9) [reply src/shared.js:858]: This is correct. I'm not sure if this is extra or needed. There are places in our code where I followed the approach of parsing the JSON data assuming no bad status codes are returned. In your opinion is this error handling necessary?

File: src/tubeValidation.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (anthonypetersen) [reply src/tubeValidation.js:40]: those variables are used in the tube definitions, so they need to be modified appropriately

File: src/shared.js [hunk 6]
```diff
@@ -834,21 +834,31 @@ export const removeBag = async(boxId, bags) => {
...
+export const searchSpecimenInstitute = async() => { // rename to searchSpecimenInstitute
     const idToken = await getIdToken();
     const response = await fetch(`${api}api=searchSpecimen`, {
         method: "GET",
         headers: {
             Authorization:"Bearer "+idToken
         }
     });
-    let a = await response.json();
+    return response.json();
```
Comment (we-ai) [inline src/shared.js:848]: I suggest handling '400' and '404' errors before a return.

File: src/events.js [hunk 7]
```diff
@@ -3782,4 +3826,47 @@ const findScannedIdInBoxesNotShippedObject = (getAllBoxesWithoutConversionRespon
...
+ *  @param {string} currTube - current specimen tube id to filter searchSpecimenInstituteList - Ex. [CXA321789 0001]
+ *  @returns {array} Example array - ['Hemolysis present']
+ *   
+*/ 
+export const getSpecimenDeviation = (searchSpecimenInstituteList, currTube) => {
+    const specimenInstituteList = searchSpecimenInstituteList;
+    const [collectionId, tubeId] = currTube.split(" ");
+    const tubeIdDeviationReasonList = deviationReasons;
+    
+    // specimenObjList - filter and return specimen object with matching tube collectedId
```
Comment (we-ai) [inline src/events.js:3832]:
```suggestion
```

File: src/pages/shipping.js [hunk 8]
```diff
@@ -248,23 +248,24 @@ export const startShipping = async (userName) => {
...
+        boxIdAndBagsObj[box['132929440']] = box['bags'];
     }
-    let currInstitute = currBox.siteAcronym;
-    let currLocation = locationConceptIDToLocationMap[currBox['560975149']]["siteSpecificLocation"];
-    let currContactInfo = locationConceptIDToLocationMap[currBox['560975149']]["contactInfo"][currInstitute];
+    const currInstitute = currBox.siteAcronym;
+    const currLocation = locationConceptIDToLocationMap[currBox['560975149']]["siteSpecificLocation"];
+    const currContactInfo = locationConceptIDToLocationMap[currBox['560975149']]["contactInfo"][currInstitute];
+    const searchSpecimenInstituteListResponse = await searchSpecimenInstitute();
+    const searchSpecimenInstituteList = searchSpecimenInstituteListResponse.data ? searchSpecimenInstituteListResponse.data : [];
```
Comment (we-ai) [inline src/pages/shipping.js:266]:
```suggestion
    const searchSpecimenInstituteList = searchSpecimenInstituteListResponse.data ?? [];
```

File: src/events.js [hunk 9]
```diff
@@ -3782,4 +3826,47 @@ const findScannedIdInBoxesNotShippedObject = (getAllBoxesWithoutConversionRespon
...
+        if (currSpecimenKey[scannedId] === currTube && currSpecimenKey[isCollected] === conceptIds.yes && currSpecimenKey[isDeviated] === conceptIds.yes) { 
+            // deviationObj - current tube's deviation keys with yes or no values 
+            const deviationObj = currSpecimenKey[deviation];
+            if (deviationObj) {
+                for (const deviation in deviationObj) {
+                    // deviation not found in the the refused shipping deviation list and deviation exists
+                    if (!refusedShippingDeviationConceptList.includes(parseInt(deviation)) && deviationObj[deviation] === conceptIds.yes) {
+                        // destructure assignment to get single deviation object match 
+                        const [filteredDeviationObj] = tubeIdDeviationReasonList.filter(deviationReason => (deviationReason['concept'] === parseInt(deviation)))
+                        const deviationLabel = filteredDeviationObj['label'];
```
[the remark this answers is not in the document]
Comment (Gbarra9) [reply src/events.js:3852]: `acceptedDeviationArr.push(tubeIdDeviationReasonList.filter(deviationReason => deviationReason['concept'] === parseInt(deviation))?.['label']);`

This should work for the more simplified approach and I don't need to use destructuring assignment here

**Pull request merged**